Compose command for collections

alzeih opened the pull request:
This allows creating files in collections, including posts and drafts.

`jekyll compose "My New Post"`
`jekyll compose "My New Post" --post`

`jekyll compose "My new draft" --draft`

`jekyll compose "My New Thing" --collection "things"`

Default front matter can be specified for each collection.

```
jekyll_compose:
  default_front_matter:
    drafts:
      layout: post
      category:
      tags:
    posts:
      published: false
      sitemap: false
    things:
      layout: thing
```

Closes #84

alzeih commented:
This PR is related to #84

ashmaroli commented:
Not so sure about what this would mean for the invocation `jekyll post` and `jekyll draft`.

Also, `drafts` and `posts` are part of the same "collection" (which is `posts`). There are no tests included in this PR, so can't tell with certainty if things work as expected..

alzeih commented:
So, eventually the `jekyll compose` command could replace `jekyll post` and `jekyll draft` as `jekyll compose` creates both posts and drafts.

I didn't want to remove `jekyll post` or `jekyll draft` with this PR because I thought those commands might first become deprecated, and then removed in a later version, if the unified `jekyll compose` was accepted.

`jekyll publish` and `jekyll unpublish` on the other hand are currently very posts collection specific because they move files between `_posts` and `_drafts`.

If collections are now to be supported, then publish and unpublish could also be made aware of collections, and instead use the yaml front matter to publish and unpublish collection items. Alternatively, publish and unpublish could just return an error and not act on collection items.

Since @ashmaroli says drafts are not a true collection, the use of `jekyll compose --collection drafts` could become confusing as a default syntax. I could add flags `jekyll compose --post`and  `jekyll compose --draft` that alias `jekyll compose --collection posts` and `jekyll compose --collection drafts` maybe?

alzeih commented:
@ashmaroli I have added tests for `jekyll compose`.

The tests are from `spec/post_spec.rb` and `spec/draft_spec.rb`, and duplicated to test the different arguments and options. There's some small changes to the tests for the different args and options for the commands, and an extra test related to this.

The only major change from the other tests was the structure of default yaml front matter, because I expect it in a different structure, so

```
jekyll_compose:
  draft_default_front_matter:
    description:
  post_default_front_matter:
    description:
```

becomes

```
jekyll_compose:
  default_front_matter:
    drafts:
      description:
    posts:
      description:
    things:
      description:
```

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,111 @@
+
+      class ComposeCommandArgParser < Compose::ArgParser
+        def validate!
+          if [options["collection"], options["post"], options["draft"]].reject(&:nil?).length > 1
```
```suggestion
          if options.values_at("collection", "draft", "post").compact.length > 1
```

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,111 @@
+        end
+
+        def date
+          options["date"].nil? ? Time.now : Date.parse(options["date"])
```
I understand that this is a copy-paste from existing commands, but IMO, there's no need to check for `nil` explicitly.
Instead check for *truthy* (neither `nil` nor `false`):
```suggestion
          options["date"] ? Date.parse(options["date"]) : Time.now
```

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,111 @@
+          else
+            "posts"
+          end
+        end
```
Instead of *negative nil-checks*, I wonder if we could just use binary operations. For example:
```ruby
options["collection"] || options["post"] || options["draft"] || "posts"
```

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,111 @@
+          else
+            "file"
+          end
+        end
```
Compact this to:
```ruby
        def resource_type
          case @params.collection
          when "posts"  then "post"
          when "drafts" then "draft"
          else
            "file"
          end
        end
```

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,111 @@
+          default_front_matter = params.config.dig("jekyll_compose", "default_front_matter", @params.collection)
+          custom_front_matter.merge!(default_front_matter) if default_front_matter.is_a?(Hash)
+
+          super({ "date" => _time_stamp }.merge(custom_front_matter))
```
```suggestion
          super({ "date" => _time_stamp }.merge!(custom_front_matter))
```

ashmaroli commented on spec/compose_spec.rb:
```diff
@@ -0,0 +1,1094 @@
+
+    context "when the post already exists" do
+      let(:name) { "An existing post" }
+      let(:filename) { "#{Time.now.strftime(Jekyll::Compose::DEFAULT_DATESTAMP_FORMAT)}-an-existing-post.md" }
```
Let us keep the line-length minimal:
```suggestion
      let(:filename) do
        "#{Time.now.strftime(Jekyll::Compose::DEFAULT_DATESTAMP_FORMAT)}-an-existing-post.md"
      end
```

ashmaroli commented:
> The only major change from the other tests was the structure of default yaml front matter

Please use the existing YAML front matter structure as currently documented in the `master` branch. Changing the structure (affecting existing commands and usage consistently) should be in a separate PR.

alzeih commented:
> > The only major change from the other tests was the structure of default yaml front matter
> 
> Please use the existing YAML front matter structure as currently documented in the `master` branch. Changing the structure (affecting existing commands and usage consistently) should be in a separate PR.

I have created PR #104 for this.

I believe I have addressed all the changes requested, and will re-request review. I rebased and squashed this PR to incorporate all the changes. Thank you for the review!

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,108 @@
+        end
+
+        def collection
+          options["collection"] || (options["post"] ? "posts" : nil) || (options["draft"] ? "drafts" : nil) || "posts"
```
Whoa! This is not equivalent to the previous implementation.
If this is the route you intend to follow, please use conventional `if .. elsif` syntax.

alzeih replied:
This is the change you requested https://github.com/jekyll/jekyll-compose/pull/103#discussion_r321104930

alzeih replied:
This was the previous implementation, which I see doesn't line up with "negative nil-checks". I think the comment may not have been attached to the right code there. 

What change did you want, if any, for this?

```
      class ComposeCommandFileInfo < Compose::FileInfo
        def resource_type
          case @params.collection
          when "posts"
            "post"
          when "drafts"
            "draft"
          else
            "file"
          end
        end
```

ashmaroli replied:
I *suggested* a simpler version:
```ruby
options["collection"] || options["post"] || options["draft"] || "posts"
```

If that wont work, then prefer using conventional syntax for readability.

alzeih replied:
`options["post"]` is a flag (no argument) and so is a value of `true`, not `"posts"`, thus the need for the extra code. I'll revert the change.

ashmaroli commented on README.md:
```diff
@@ -39,10 +40,20 @@ Create your new page using:
+    # or by using the compose command with post specified
+    $ bundle exec jekyll compose "My New Post" --post
+    # or by using the compose command with the posts collection specified
+    $ bundle exec jekyll compose "My New Post" --collection "posts"
```
Since these commands are not part of a workflow, let us split this into separate individual examples. Also, using syntax-highlighting will further delineate comments:
```sh
    # or by using the compose command
    $ bundle exec jekyll compose "My New Post"
```
```sh
    # or by using the compose command with post specified
    $ bundle exec jekyll compose "My New Post" --post
```
```sh
    # or by using the compose command with the posts collection specified
    $ bundle exec jekyll compose "My New Post" --collection "posts"
```

ashmaroli commented on README.md:
```diff
@@ -39,10 +40,20 @@ Create your new page using:
+    # or by using the compose command with draft specified
+    $ bundle exec jekyll compose "My new draft" --draft
+    # or by using the compose command with the drafts collection specified
+    $ bundle exec jekyll compose "My new draft" --collection "drafts"
```
Same as above..

ashmaroli commented on lib/jekyll/commands/compose.rb:
```diff
@@ -0,0 +1,108 @@
+    class ComposeCommand < Command
+      def self.init_with_program(prog)
+        prog.command(:compose) do |c|
+          c.syntax "compose [--post|--draft|--collection COLLECTION] NAME"
```
Was this there in the previous implementation?
Either ways, please remove the *options* from the syntax. That is going to confuse users.

```suggestion
          c.syntax "compose NAME"
```
This should default to using `--post` if no options were provided.

ashmaroli commented:
@alzeih I took the liberty of pushing further changes myself instead of asking you to do it.
You can update your local version of this branch by running the following:
```
git pull origin patch-9
```

alzeih commented:
Rebased for conflicts with master branch `.rubocop_todo.yml`

alzeih commented:
@DirtyF I chose name over filename due to how `jekyll post` and `jekyll draft` worked. I thought it had the advantage of the auto slugging the title and populating the yaml front matter.

What features do you see `jekyll compose _things/my-new-thing.md` do above say `$EDITOR _things/my-new-thing.md` ? (I haven't used Hugo or know what you like about it)

DirtyF commented:
@alzeih Not a problem. It's just a _personal_ preference, mainly because I find it way easy to type and memorize.

Hugo has [archetypes](https://gohugo.io/content-management/archetypes/) (it's a whole different approach to provide similar feature)

alzeih commented:
@DirtyF I can see the advantage of having automatically selected and prepopulated templates for collection items, including content! I'll think about how it might be done, though probably not for this PR.

alzeih commented:
This PR is marked as approved, but has not been merged. Are there any outstanding issues I can address?

ashmaroli commented:
@alzeih This PR will be merged after 1 more approval from @jekyll/plugin-core

DirtyF commented:
@jekyllbot: merge +minor
